### ISSUE TITLE
Files not being deleted from image when using docker with userns and overlay2 storage

### ISSUE THREAD

**Issue by karlh1** [moby/moby#32919]
**Description**

I'm using dockerd with userns-remap and the overlay2 storage driver enabled and noticed some strange behavior while testing my first images. Files that should have been deleted during the build phase are still present when running the image. 

Dockerfile:
```
FROM debian:jessie

COPY testfile /testfile
RUN rm /testfile
CMD ["/bin/bash"]
```

daemon.json:
```
{
    "storage-driver": "overlay2",
    "userns-remap": "tim"
}
```

/etc/subuid and /etc/subgid:
```
systemd-timesync:100000:65536
systemd-network:165536:65536
systemd-resolve:231072:65536
systemd-bus-proxy:296608:65536
tim:362144:65536
sshd:427680:65536
```

**Steps to reproduce the issue:**
```
$ docker build -t testimage .
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM debian:jessie
jessie: Pulling from library/debian
cd0a524342ef: Pull complete 
Digest: sha256:c3f000ba6bbe71906ca249be92bd04dc3f514d2dd905e0c5f226e8035ee55031
Status: Downloaded newer image for debian:jessie
 ---> 054abe38b1e6
Step 2/4 : COPY testfile /testfile
 ---> fdf736b5c7ee
Removing intermediate container 36d416aaf754
Step 3/4 : RUN rm /testfile
 ---> Running in ebc0dd7c511f
 ---> 2326dae61693
Removing intermediate container ebc0dd7c511f
Step 4/4 : CMD /bin/bash
 ---> Running in 8891670579f1
 ---> ba290ae80cc6
Removing intermediate container 8891670579f1
Successfully built ba290ae80cc6
Successfully tagged testimage:latest

$ docker run -it --rm testimage
```

**Describe the results you received:**
```
root@61d866afa313:/# ls
bin   dev  home  lib64  mnt  proc  run   srv  testfile  usr
boot  etc  lib   media  opt  root  sbin  sys  tmp       var
```
**Describe the results you expected:**
testfile shouldn't be present.

**Additional information you deem important (e.g. issue happens only occasionally):**
When I ran the same tests using the aufs storage driver the file wasn't present as expected. Turing off userns-remap and using overlay2 also produces the correct result.

**Output of `docker version`:**

```
Client:
 Version:      17.05.0-ce-rc2
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   c57fdb2
 Built:        Thu Apr 27 03:42:00 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce-rc2
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c57fdb2
 Built:        Thu Apr 27 03:42:00 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

The current stable version has the same issue as well.

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 17.05.0-ce-rc2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
 userns
Kernel Version: 4.9.0-2-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.6MiB
Name: test-debian
ID: T7RU:BNET:J4IB:PG4G:LPLX:DPQC:EBQK:KTVH:EL2J:RKPI:UBSL:MN2Y
Docker Root Dir: /var/lib/docker/362144.362144
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Tested everything inside a VirtualBox VM running Debian Stretch (amd64) with only minimal changes (using sysvinit instead of systemd).

**Comment by F30** [comment]
I can confirm this in a rather similar setting:

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 22
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.13.0-0.bpo.1-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.957GiB
Name: workshop
ID: DUGQ:NVYJ:6DZ4:ZN6H:MQLJ:MNV6:PMMJ:H42A:YLBH:PMGW:JIZ6:DIHP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

The issue occurs with kernel version 4.9.0-4-amd64 as well as 4.13.0-0.bpo.1-amd64. Disabling user namespaces fixes it.

**Comment by F30** [comment]
As mentioned in #35709, this appears fixed in Docker 18.03.

**Comment by thaJeztah** [comment]
Ah, yes, this should be resolved through https://github.com/moby/moby/pull/35822. Let me close this issue, but feel free to comment if you still run into this
